src/task/syncBorrowedShips.ts

```diff
@@ -0,0 +1,190 @@
+      this.nesting.push(next);
+      it = next;
+    }
+  }
+
+  get outermost() {
+    return this.nesting[this.nesting.length - 1];
+  }
+
+  get asset() {
+    return this.nesting[0];
+  }
+
+  get inStation() {
+    return this.outermost.locationType === 'station';
+  }
+
+  get inPlayerStructure() {
+    return (
+      this.outermost.locationType === 'item' &&
+      this.outermost.locationFlag === 'Hangar'
+    );
+  }
+
+  get insideCorpShip() {
+    return this.nesting.slice(1).some((asset) => isCorpShip(asset));
+  }
+
+  describeLocation(locCache: LocationCache) {
+    let tokens = [];
+    for (let i = 1; i < this.nesting.length; ++i) {
+      const n = this.nesting[i];
+      if (!n.name) continue;
+      // Currently, if something has a name, then it's a ship, and its useful
+      // to include its type and the "position" (e.g. fleet hangar) where the
+      // asset is located. May be less useful for other named containers, if
+      // we start fetching names for other assets.
+      const flag = formatLocationFlag(this.nesting[i - 1].locationFlag);
+      tokens.push(`${n.name} (${n.typeName}), ${flag}`);
+    }
+    const last = locCache.get(this.outermost.locationId);
+    if (last !== undefined) {
+      tokens.push(last);
+    }
+    if (!tokens.length) {
+      return formatLocationFlag(this.asset.locationFlag);
+    }
+    tokens.reverse();
+    return tokens.join(' > ');
+  }
+}
+
+/**
+ * Given the list of assets of a single character, finds the ships that look
+ * like they might belong to the corp. To provide a useful description of the
+ * locations of those ships, may need to retrieve station/structure names from
+ * ESI. Uses location cache for that.
+ */
+async function findShips(
+  characterId: number,
+  token: string,
+  assets: Asset[],
+  locCache: LocationCache,
+): Promise<CharacterShipRow[]> {
+  const assetMap = arrayToMap(assets, 'itemId');
+  const ships = assets
+    .filter((asset) => isCorpShip(asset))
+    .map((asset) => new NestedAsset(asset, assetMap))
+    // No need to be noisy about a corp ship that's inside another
+    // corp ship that we'll report anyway.
+    .filter((asset) => !asset.insideCorpShip);
+
+  const stationIds = new Set<number>();
+  for (let s of ships) {
+    if (s.inStation) {
+      stationIds.add(s.outermost.locationId);
+      continue;
+    }
+    if (s.inPlayerStructure) {
+      await locCache.cachePlayerStructure(s.outermost.locationId, token);
+    }
+  }
+  await locCache.cacheStations(Array.from(stationIds));
+
+  return ships.map(
+    (it) =>
+      <CharacterShipRow>{
+        characterId: characterId,
+        itemId: it.asset.itemId,
+        typeId: it.asset.typeId,
+        name: it.asset.name!,
+        locationDescription: it.describeLocation(locCache),
+      },
+  );
+}
+
+async function executor(db: Tnex, job: JobLogger) {
+  job.setProgress(0, undefined);
+  const characterIds = await dao.roster.getCharacterIdsOwnedByMemberAccounts(
```

Comment:
Would it make sense to track the last time we did this check for any given character, and only re-fetch if it's at least N minutes stale? That'd avoid us doing unnecessary work in the case where the task gets interrupted and restarted.

Reply:
Done